[RVP] add p extension spec 0.94 support

linsinan1995 opened the pull request:
- Andes-tech and Shao-Chung Wang implement the RISC-V 'P' extension with the spec version 0.51 in GCC 7.6
- Sinan Lin update the implementation to the spec version ~~0.93~~ 0.94 with GCC 10.2
- ~~Add commit 0b7b4710111e and 4b81528241c for fixing configure option parsing error~~

kito-cheng commented:
- This patch set is too huge, could you split it into several patches? I guess at list you could split into following different parts:
  - MD pattern changes
    - Minimal support for new mode, e.g. move pattern + necessary hook changes in riscv.c
    - standard pattern for those modes.
    - other MD patterns, at least one patch for each sub-extension to make it easier to review.
  - Intrinsic/built-in functions support
  - Auto vectorization support
- Don't merge other standalone commit to your commit, like 0b7b471 and 4b81528, that will cause author info is gone, and also impossible to distinguish which part is changed by you.

linsinan1995 commented:
Hi @kito-cheng , thanks for the suggestions! I will do that later on. Also, I wonder can we have a new experimental branch for p ext, since it might be still huge even if we split it into small pieces.

linsinan1995 committed:
[MD] add inst: add* kadd* ukadd* sub* ksub* uksub* kabs*

linsinan1995 committed:
[MD] add inst: ave,clorv*,clrs,clz,bitrev*,cmpeq*,cmix
- constraints and predicate for unsigned 5 and 6 bit imm and register

linsinan1995 committed:
[MD] add inst: *cras* and *crsa* instructions

linsinan1995 committed:
[MD] add inst: insb, KDMBB, KDMBT, KDMTT, KDMABB, KDMABT, KDMATT, KHMBB, KHMBT, KHMTT, KMABB, KMABT, KMATT, KHM8, KHMX8, KHM16, KHMX16
- Add integer constant and bit pos in constraints and predicates

linsinan1995 committed:
[MD] add inst: KMADA, KMAXDA, KMADS, KMADRS, KMAXDS, KMAR64, KMMAC, KMMAC.u, KMMAWB, KMMAWB.u, KMMAWB2, KMMAWB2.u, KMMAWT2, KMMAWT2.u, KMMWB2, KMMWB2.u, KMMWT2, KMMWT2.u, KMSR64
- add 15, 16 integer constant in predicate and constraints

linsinan1995 committed:
[MD] add inst: KSLL[i]W, KSLL[i]8, KSLL[i]16, KSLL[i]32, KSLRA8[.u], KSLRA16[.u], KSLRAW[.u]
- add constraints and predicates for integer constant 3, 4

linsinan1995 committed:
[MD] add inst: kstas16, kstas32, kstsa16, kstsa32

linsinan1995 committed:
[MD] add inst: KWMMUL[.u], MADDR32, MSUBR32, MULR64,
MULSR64, ZBPBO: MIN, MAX

linsinan1995 committed:
[MD] add inst: PBSAD, PBSADA, PKBB[16|32], PKBT[16|32], PKTT[16|32], PKTB[16|32] ZBPBO: pack, packu
- Add predicate for integer constant 1, 2

linsinan1995 committed:
[MD] add inst: [U]RADD[8|16|32|64|W], [U]RSUB[8|16|32|64|W]

linsinan1995 committed:
[MD] add inst: RDOV, RSTAS[16|32], RSTSA[16|32]

linsinan1995 committed:
[MD] add inst: SCLIP8, SCLIP16, SCLIP32, SCMPLE8, SCMPLE16, SCMPLT8, SCMPLT16

linsinan1995 committed:
[MD] add inst: SMALBB, SMALBT, SMALTT, SMAL

linsinan1995 committed:
[MD] add inst: SMALDA, SMALXDA, SMALDS, SMALDRS, SMALXDS

linsinan1995 committed:
[MD] add inst: SMAR64, UMAR64, [U]SMAX 8|16  and [U]SMIN 8|16

linsinan1995 committed:
[MD] add inst: SMAQA, SMAQA.SU, UMAQA, UMAQA.SU

linsinan1995 committed:
[MD] add inst: SMBB 16|32, SMBT 16|32, SMTT 16|32
- Add predicate and contraint for 2 bit imm

linsinan1995 committed:
[MD] add inst: SMDS, SMDRS, SMXDS, SMDS32, SMDRS32, SMXDS32

linsinan1995 committed:
[MD] add inst: SMMUL[.u], SMMWB[.u], SMMWT[.u]

linsinan1995 committed:
[MD] add inst: SMSLDA, SMSLXDA, SMSR64, UMSR64

linsinan1995 committed:
[MD] add inst: SMUL 8|16, SMULX 8|16, UMUL 8|16, UMULX 8|16

linsinan1995 committed:
[MD] add inst: SRA[I] 8|16|32, SRL[I] 8|16, SRA[I] 8|16|32 .u, SRL[I] 8|16|32 .u SLL[I] 8|16|32

linsinan1995 committed:
[MD] add inst: STAS 16|32, STSA 16|32

linsinan1995 committed:
[MD] add inst: SUNPKD810, SUNPKD820, SUNPKD830, SUNPKD831, SUNPKD832, ZUNPKD810, ZUNPKD820, ZUNPKD830, ZUNPKD831, ZUNPKD832

linsinan1995 committed:
[MD] add inst: swap8, rev8.h

- swap16 is an alias of PKBT16 Rd, Rs1, Rs1 from spec v 0.93

linsinan1995 committed:
[MD] add inst: UCLIP8|16|32, UCMPLE8|16, UCMPLT8|16

linsinan1995 committed:
[MD] add inst: ukmar64, ukmsr64, ukstas16|32, UKSTSA16|32

linsinan1995 committed:
[MD] add inst: WEXTI, WEXT, URSTSA 16|32, URSTAS 16|32

linsinan1995 committed:
[MD] add inst: KDMBB16, KDMBT16, KDMTT16, KDMABB16, KDMABT16, KDMATT16

linsinan1995 committed:
[MD] add inst: KHMBB16, KHMBT16, KHMTT16, KMABB32, KMABT32, KMATT32, KMADA32, KMAXDA32, KMDA32, KMXDA32

linsinan1995 committed:
[builtin] Add type conversion while expanding builtins

linsinan1995 committed:
[Hook] Add TARGET_VECTOR_MODE_SUPPORTED_P implementation

linsinan1995 committed:
[Hook] Add even-odd pair of register support in TARGET_HARD_REGNO_MODE_OK hook

yetingk reviewed (commented):
For [[Hook] Add TARGET_VECTOR_MODE_SUPPORTED_P implementation](https://github.com/riscv/riscv-gcc/pull/258/commits/0f432e3385ed7c7133d3a6bad71c0565e1b54cdc
)
Since some instructions like `smul16` in rv32p need vector register with 64-bit, I think you should also consider `MODE_VECTOR_INT`.

linsinan1995 commented:
@fakepaper56 Thanks, fixed in new [commit](https://github.com/riscv/riscv-gcc/pull/258/commits/316b3e7f576e76b616cd36e8e2440c0ca0a42f6a).

kito-cheng commented:
Could you rebase and fix the conflict?

linsinan1995 commented:
No conflict now.

kito-cheng commented:
Oops, I still saw `This branch cannot be rebased due to conflicts` on github?

kito-cheng commented:
Oh, but it could be merged by `merge` or `squash and merge`

kito-cheng commented on gcc/config/riscv/rvp.md:
```diff
@@ -243,3 +254,147 @@
+
+(define_insn "riscv_<rvp_optab>si2"
+  [(set (match_operand:SI 0 "register_operand" "=r")
+        (unop:SI (match_operand:SI 1 "register_operand" "r")))]
```
I would suggest using `standard pattern name`[1] for those pattern, I mean without the riscv_ prefix, using `standard pattern name` can teach GCC to generate that instruction when user using `__builtin_clz` and `__builtin_clrsb`[2]

[1] https://gcc.gnu.org/onlinedocs/gccint/Standard-Names.html#index-clrsbm2-instruction-pattern
[2] https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005fclz

linsinan1995 replied:
Thanks for your info! Fixed in commit 715c9ca. Also, there are many other patterns that can be renamed (remove riscv_ prefix), such as smin, smax, etc. Maybe I can just remove the `riscv_` prefix in all the names?

kito-cheng reviewed (commented):
This changes are so huge, I guess I need take long time to review that, and I'll submit my review when I review part of patch.

linsinan1995 commented on gcc/config/riscv/rvp.md:
```diff
@@ -243,3 +254,147 @@
+  "TARGET_ZPN && !TARGET_64BIT"
+  "ave\t%0, %1, %2"
+  [(set_attr "type" "dsp")
+   (set_attr "mode" "SI")])
```
fixed in the new commit eb5b39a

linsinan1995 commented on gcc/config/riscv/riscv.md:
```diff
@@ -56,6 +56,10 @@
   UNSPEC_UKADDH
   UNSPEC_UKSUBH
+  UNSPEC_BITREV
+  UNSPEC_CLROV
```
fixed in the new commit eb5b39a

linsinan1995 commented on gcc/config/riscv/rvp.md:
```diff
@@ -0,0 +1,7133 @@
+(define_code_attr ph [(plus "") (minus "")])
+
+;; bit count, clz and clrs
+(define_code_iterator unop [clrsb clz])
```
fixed in the new commit eb5b39a

linsinan1995 commented on gcc/config/riscv/rvp_intrinsic.h:
```diff
@@ -0,0 +1,43 @@
+typedef unsigned int uint32x2_t __attribute__((vector_size(8)));
+typedef unsigned int uint32x4_t __attribute__((vector_size(16)));
+
+#define __riscv__kabsw(a) \
```
it only helps intrinsic header more readable, e.g. 
```c
#if defined(__riscv_zpsf) && __riscv_xlen == 64
#define __rv__v_smar64 __riscv__v_smar64
#define __rv__v_smsr64 __riscv__v_smsr64
...
```
instead of several lines for one intrinsic. I can remove them if it is not appropriate.

linsinan1995 commented on gcc/config/riscv/predicates.md:
```diff
@@ -212,3 +212,21 @@
   return riscv_gpr_save_operation_p (op);
 })
+
+(define_predicate "register_even_operand"
```
I think it is not necessary, and it was removed by commit ecd8198

kito-cheng replied:
Could you squash your patch together? I mean merge the fixup patch into the original one, since I am review the commit one-by-one, and it would be easier review when you submit patch to upstream.

linsinan1995 replied:
Hi Kito, I have sorted them up.

linsinan1995 committed:
Initial support for P extension with spec 0.93

- initial support for p ext with spec 0.93
- add instruction kabsw

kito-cheng commented on gcc/config/riscv/rvp.md:
```diff
@@ -17,11 +17,211 @@
+  [(set_attr "type" "dsp64")
+   (set_attr "mode" "DI")])
+
+;; uk|k add|sub w|h
```
Could you simplify those pattern by `define_int_iterator` and `define_int_attr`?

kito-cheng commented on gcc/config/riscv/rvp.md:
```diff
@@ -1140,4 +1143,535 @@
+   (set_attr "mode" "V8QI")])
+
+;; KDMBB, KDMBT, KDMTT (32 (*) (xlen, xlen) for vector mode & 32 (*) (32, 32) for non-vector mode)
+(define_insn "kdmbb<mode>"
```
`kdm*<mode>` seems can be merge with `kdm*` and `v64_kdm*` pattern?

kito-cheng commented on gcc/config/riscv/rvp.md:
```diff
@@ -2814,3 +2814,107 @@
+(define_insn "ksll"
+  [(set (match_operand:SI 0 "register_operand"               "=   r, r")
+	(ss_ashift:SI (match_operand:SI 1 "register_operand" "    r, r")
+		      (match_operand:SI 2 "rimm5u_operand"   " Iu05, r")))]
```
`u05` is enough here.

kito-cheng commented on gcc/config/riscv/rvp.md:
```diff
@@ -3053,3 +3053,115 @@
+   (set_attr "mode" "SI")])
+
+;; MULR64, MULSR64
+(define_insn "rvp_umulsidi3"
```
`umulsidi3` is a standard pattern name, so I would suggest remove `rvp_`.

kito-cheng commented on gcc/config/riscv/rvp.md:
```diff
@@ -3053,3 +3053,115 @@
+  [(set_attr "type"   "dsp")
+   (set_attr "mode"   "DI")])
+
+(define_insn "rvp_mulsidi3"
```
`mulsidi3`  is a standard pattern name, so I would suggest remove `rvp_` here too.

linsinan1995 replied:
Thanks for the suggestions. 

The `mulsidi3` and `umulsidi3` are defined in riscv.md in 32 bit, so I put 
```
if (TARGET_ZPN)
  emit_insn (gen_rvp_<u>mulsidi3 (operands[0], operands[1], operands[2]));
else
  ...
```
under `define_expand "<u>mulsidi3"` and define `<u>mulsidi3`  in `rvp.md` for 64 bit. Does it sound a good solution?

kito-cheng replied:
Yeah, that seems right solution :)

linsinan1995 commented:
Hi Kito, the feedback problems above are addressed, except removing the unspec of vector compare pattern, which I will investigate later.

marcfedorow commented on gcc/config.gcc:
```diff
@@ -525,6 +525,7 @@ pru-*-*)
 	;;
 riscv*)
 	cpu_type=riscv
+	extra_headers="rvp_intrinsic.h"
```
I suggest `rvpintrin.h` similar to `rvkintrin.h`

marcfedorow commented on gcc/config/riscv/constraints.md:
```diff
@@ -81,3 +81,73 @@
+       (match_test "ival < (1 << 6) && ival >= 0")))
+
+(define_constraint "C00"
+  "Constant value 1"
```
0?

marcfedorow commented on gcc/common/config/riscv/riscv-common.c:
```diff
@@ -68,6 +68,8 @@ riscv_implied_info_t riscv_implied_info[] =
   {"zks", "zkg"},
   {"zks", "zkb"},
+  {"p", "zpn"},
+  {"p", "zpsf"},
```
zprv?

linsinan1995 replied:
The mapping here means the `p` flag in `--with-arch=rv{xlen}gcp` or `-march=rv{xlen}gcp` can be expanded to `rv{xlen}gc_zpn_zpsf`.

marcfedorow replied:
I agree. Still I think that zprv should appear on xlen=64.
When I run `riscv64-unknown-elf-gcc -march=rv64imcbp -mabi=lp64 .....` on compiler with this PR, I get
```
rv64up.c: warning: implicit declaration of function '__rv__kadd32' [-Wimplicit-function-declaration]
    |   if (0x0000000000000000 != __rv__kadd32(0x0000000000000000, 0x0000000000000000)) return 536;
    |                             ^~~~~~~~~~~~
```

This is so because `rv64imcbp` expands to `rv64imcb_zpn_zpsf` whilest it should expand to `rv64imcb_zpn_zpsf_zprv` because `zprv` is required on `xlen=64` by spec (I simplify a bit: I didn't expand b to its zeds here because it is not the issue).
If I do not treat warnings as errors, it obviously fails on linking with:
```
objs/rv64imcb@pcsh-bm1_rv64up/rv64up.o: in function `.L0 ': undefined reference to `__rv__kadd32'
```
I faced a more general problem with this PR: when I compile my code with `riscv64-unknown-elf-gcc -march=rv64imcb -mabi=lp64 .....` (i.e. **without** p-ext) it fails on linking with this error:
```
objs/rv64imcb@pcsh-bm1_rv64up/rv64up.o: in function `.L0 ': undefined reference to `__builtin_riscv_kadd16'
```

I have my own p-ext intrinsic file (which is generally
```cpp
#ifdef __riscv_zpn
static inline __rv__insn(uintXLEN_t rs1, uintXLEN_t rs2) { uintXLEN_t rd; __asm__("insn %0, %1, %2" : "=r"(rd) : "r"(rs1), "r"(rs2)); return rd; }
#else
static inline __rv__insn(uintXLEN_t rs1, uintXLEN_t rs2) { SOME_GENERIC_CODE; return rd; }
#endif
```
with different variations of types, number of arguments, clobbers and other stuff). I did so because I am not a pro at machine description language and also it seems to be a common way to implement intrinsics (e.g. https://github.com/rvkrypto/rvkrypto-fips/blob/main/rvkintrin.h).

So my intrinsics produce an assembler code with the specified insn if p-ext is enabled and make a call to (or inline -- depends on optimization levels) some generic-code functions otherwise.

When I found this PR, I expected it works the same way and in addition allows to optimize some generic code with emitting of p-ext insns.
As you mentioned in another topic, "generic code" should use very specific types defined in the same file where intrinsics are -- so we barely expect that such insns will appear on some code with common scalar types as b-ext insns do. It is ok for me: I understand that b-ext machine descriptions implement some simple stuff (e.g. bit set) and this PR operates with packed types so this is either very hard or impossible to map a set of scalar type insns to p-ext insn in most cases. Still I don't understand why intrinsics do not work (I mean do not provide generic code) without enabled p-ext and what should be done if I want to compile some file with the compiler which includes this intrinsic file without p-ext insns.

linsinan1995 replied:
Hi @marcfedorow,

> This is so because rv64imcbp expands to rv64imcb_zpn_zpsf whilest it should expand to rv64imcb_zpn_zpsf_zprv because zprv is required on xlen=64 by spec (I simplify a bit: I didn't expand b to its zeds here because it is not the issue).

Thank you for pointing it out. I add a fix to support a different expansion on p flag (rv32p will be expanded into rv32_zpn_zpsf, and rv64p will be rv64_zpn_zprv_zpsf).

> with different variations of types, number of arguments, clobbers and other stuff). I did so because I am not a pro at machine description language and also it seems to be a common way to implement intrinsics (e.g. https://github.com/rvkrypto/rvkrypto-fips/blob/main/rvkintrin.h).
So my intrinsics produce an assembler code with the specified insn if p-ext is enabled and make a call to (or inline -- depends on optimization levels) some generic-code functions otherwise.

I think we do not have such a plan so far, since it requires a great amount of workload. I think it is a good idea though, so I will forward your message to other PLCT members in the further meetup. 

> As you mentioned in another topic, "generic code" should use very specific types defined in the same file where intrinsics are -- so we barely expect that such insns will appear on some code with common scalar types as b-ext insns do. It is ok for me: I understand that b-ext machine descriptions implement some simple stuff (e.g. bit set) and this PR operates with packed types so this is either very hard or impossible to map a set of scalar type insns to p-ext insn in most cases. Still I don't understand why intrinsics do not work (I mean do not provide generic code) without enabled p-ext and what should be done if I want to compile some file with the compiler which includes this intrinsic file without p-ext insns.

Vector type `int16x4_t`  doesn't need to be from p-ext header, since it is native support in GCC (https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html). To use intrinsic might be the only solution if you try to use scalar type instead of vector type (e.g. put int16x4_t data onto int64_t field), but p-ext insn is possible to be generated from scalar code through auto-vectorization. e.g.

```c
#include <rvp_intrinsic.h>
#include <stdint.h>

typedef short v4hi __attribute__((vector_size (8)));

v4hi v_sadd16_spn (v4hi ra, v4hi rb)
{
  return ra + rb;
}

int16_t *v_sadd16_arr (int16_t *ra, int16_t *rb, int len)
{
  for (int i = 0; i < len; i++)
	  ra[i] += rb[i];
  return ra;
}
```

riscv64-unknown-elf-gcc -S -O3 add16.c
```
	.file	"add16.c"
	.option nopic
	.attribute arch, "rv64i2p0_m2p0_a2p0_f2p0_d2p0_c0p0_p2p0_zpn2p0_zprv2p0_zpsf2p0"
	.attribute unaligned_access, 0
	.attribute stack_align, 16
	.text
	.align	1
	.globl	v_sadd16_spn
	.type	v_sadd16_spn, @function
v_sadd16_spn:
	add16	a0, a0, a1
	ret
	.size	v_sadd16_spn, .-v_sadd16_spn
	.align	1
	.globl	v_sadd16_arr
	.type	v_sadd16_arr, @function
v_sadd16_arr:
	ble	a2,zero,.L4
	addi	a5,a1,2
	addiw	a3,a2,-1
	sub	a5,a0,a5
	sext.w	a4,a3
	sltiu	a5,a5,5
	li	a6,5
	xori	a5,a5,1
	sgtu	a4,a4,a6
	and	a5,a4,a5
	sext.w	t1,a2
	beq	a5,zero,.L5
	or	a5,a1,a0
	andi	a5,a5,7
	bne	a5,zero,.L5
	srliw	a7,t1,2
	slli	a7,a7,3
	mv	a5,a0
	mv	a3,a1
	add	a7,a7,a0
.L6:
	ld	a4,0(a5)
	ld	a6,0(a3)
	addi	a5,a5,8
	addi	a3,a3,8
	add16	a4, a4, a6
	sd	a4,-8(a5)
	bne	a5,a7,.L6
	andi	a5,t1,-4
	mv	a4,a5
	beq	t1,a5,.L4
	slli	a5,a5,32
	srli	a5,a5,31
	add	a7,a0,a5
	add	a3,a1,a5
	lhu	a6,0(a3)
	lhu	t1,0(a7)
	addiw	a3,a4,1
	addw	a6,a6,t1
	sh	a6,0(a7)
	bge	a3,a2,.L4
	addi	a3,a5,2
	add	a6,a0,a3
	add	a3,a1,a3
	lhu	a3,0(a3)
	lhu	a7,0(a6)
	addiw	a4,a4,2
	addw	a3,a3,a7
	sh	a3,0(a6)
	bge	a4,a2,.L4
	addi	a5,a5,4
	add	a4,a0,a5
	add	a1,a1,a5
	lhu	a5,0(a1)
	lhu	a3,0(a4)
	addw	a5,a5,a3
	sh	a5,0(a4)
	ret
.L5:
	slli	a5,a3,32
	srli	a3,a5,31
	addi	a2,a0,2
	mv	a5,a0
	add	a2,a2,a3
.L8:
	lhu	a4,0(a5)
	lhu	a3,0(a1)
	addi	a5,a5,2
	addi	a1,a1,2
	addw	a4,a4,a3
	sh	a4,-2(a5)
	bne	a5,a2,.L8
.L4:
	ret
	.size	v_sadd16_arr, .-v_sadd16_arr
	.ident	"GCC: (GNU) 10.2.0"
```

I hope I have answered all your questions. Thanks again.

marcfedorow commented on gcc/config/riscv/rvp_intrinsic.h:
```diff
@@ -0,0 +1,1578 @@
+  (__builtin_riscv_ukmsr64 ((r), (a), (b)))
+#define __rv__maddr32(a, b) \
+  (__builtin_riscv_maddr32 ((a), (b)))
+#define __rv__msubr32(a, b) \
```
r, a, b

marcfedorow commented on gcc/config/riscv/rvp_intrinsic.h:
```diff
@@ -0,0 +1,1578 @@
+  (__builtin_riscv_ukmar64 ((r), (a), (b)))
+#define __rv__ukmsr64(r, a, b) \
+  (__builtin_riscv_ukmsr64 ((r), (a), (b)))
+#define __rv__maddr32(a, b) \
```
r, a, b

marcfedorow commented:
I am unable to emit p-ext insns from generic code.
Here are my examples:

```c
#include <rvp_intrinsic.h>
#include <stdint.h>
#include <stdlib.h>

#define ADD_SUB(u, n, BODY)                                 \
    uint64_t rd = 0;                                        \
    for (int i = 0; i < __riscv_xlen; i += n) {             \
        u##int##n##_t rs1s = rs1 >> i, rs2s = rs2 >> i;     \
        rd |= (uint64_t)((BODY) & UINT##n##_MAX) << i;      \
    }                                                       \
    return rd

static inline uint64_t add16(uint64_t rs1, uint64_t rs2)
{
    ADD_SUB( , 16, rs1s + rs2s );
}

volatile long unsigned int foo(){
  return rand();
}

int main()
{
  uint64_t a = foo(), b = foo();
  return __rv__add16(a, b) != add16(b, a);
}
```
compiles to
```asm
 <main>:
1141                	addi	sp,sp,-16
e406                	sd	ra,8(sp)
e022                	sd	s0,0(sp)
66e000ef          	jal	ra,800015c8 <rand>
842a                	mv	s0,a0
668000ef          	jal	ra,800015c8 <rand>
02055693          	srli	a3,a0,0x20
02045613          	srli	a2,s0,0x20
872a                	mv	a4,a0
4105559b          	sraiw	a1,a0,0x10
4104579b          	sraiw	a5,s0,0x10
0106939b          	slliw	t2,a3,0x10
0106189b          	slliw	a7,a2,0x10
6541                	lui	a0,0x10
fff50293          	addi	t0,a0,-1 # ffff
00b7833b          	addw	t1,a5,a1
4103d81b          	sraiw	a6,t2,0x10
4108de1b          	sraiw	t3,a7,0x10
00537eb3          	and	t4,t1,t0
01c80f3b          	addw	t5,a6,t3
43075f93          	srai	t6,a4,0x30
43045593          	srai	a1,s0,0x30
0104161b          	slliw	a2,s0,0x10
0107139b          	slliw	t2,a4,0x10
005f76b3          	and	a3,t5,t0
01f5853b          	addw	a0,a1,t6
000e879b          	sext.w	a5,t4
4106531b          	sraiw	t1,a2,0x10
4103d81b          	sraiw	a6,t2,0x10
00557eb3          	and	t4,a0,t0
01079893          	slli	a7,a5,0x10
02069e13          	slli	t3,a3,0x20
01030f3b          	addw	t5,t1,a6
01c8efb3          	or	t6,a7,t3
030e9593          	slli	a1,t4,0x30
005f72b3          	and	t0,t5,t0
00bfe7b3          	or	a5,t6,a1
0002869b          	sext.w	a3,t0
00d7e533          	or	a0,a5,a3
40870477          	add16	s0,a4,s0
60a2                	ld	ra,8(sp)
40850733          	sub	a4,a0,s0
6402                	ld	s0,0(sp)
00e03533          	snez	a0,a4
0141                	addi	sp,sp,16
8082                	ret
```
and
```c
#include <rvp_intrinsic.h>
#include <stdint.h>
#include <stdlib.h>

typedef union {
  uint64_t packed;
  uint16_t arr[4];
} pack;

static inline uint64_t add16(uint64_t rs1, uint64_t rs2)
{
  pack a, b, c;
  a.packed = rs1, b.packed = rs2;
  for (short i = 0; i < 4; ++i){
    c.arr[i] = a.arr[i] + b.arr[i];
  }
  return c.packed;
}

volatile long unsigned int foo(){
  return rand();
}

int main()
{
  uint64_t a = foo(), b = foo();
  return __rv__add16(a, b) != add16(b, a);
}
```
compiles to
```asm
 <main>:
1141                	addi	sp,sp,-16
e406                	sd	ra,8(sp)
e022                	sd	s0,0(sp)
63a000ef          	jal	ra,80001594 <rand>
842a                	mv	s0,a0
634000ef          	jal	ra,80001594 <rand>
01045693          	srli	a3,s0,0x10
01055793          	srli	a5,a0,0x10
00f682bb          	addw	t0,a3,a5
02045713          	srli	a4,s0,0x20
02055613          	srli	a2,a0,0x20
00a4033b          	addw	t1,s0,a0
00c703bb          	addw	t2,a4,a2
03055813          	srli	a6,a0,0x30
03045593          	srli	a1,s0,0x30
0802c8bb          	zext.h	a7,t0
01089e13          	slli	t3,a7,0x10
00b80ebb          	addw	t4,a6,a1
08034f3b          	zext.h	t5,t1
0803cfbb          	zext.h	t6,t2
01cf67b3          	or	a5,t5,t3
020f9293          	slli	t0,t6,0x20
030e9693          	slli	a3,t4,0x30
0057e733          	or	a4,a5,t0
00d76633          	or	a2,a4,a3
40850577          	add16	a0,a0,s0
40c50433          	sub	s0,a0,a2
60a2                	ld	ra,8(sp)
00803533          	snez	a0,s0
6402                	ld	s0,0(sp)
0141                	addi	sp,sp,16
8082                	ret
```

I also tried without `volatile`, here is my foo:
```c
long unsigned int foo(){
  static int bar = 12345;
  return ^= 54321;
}
```
Results are pretty same -- generic code does not emit as an instruction.

Also I tried this:
```c
int main()
{
  uint64_t a = 12345, b = 54321;
  return __rv__add16(a, b) != add16(b, a);
}
```
It compiled to this:
```asm
c0000797          	auipc	a5,0xc0000
8e878793          	addi	a5,a5,-1816 # 40000838 <__TMC_END__+0x10>
c0000297          	auipc	t0,0xc0000
8e828293          	addi	t0,t0,-1816 # 40000840 <__TMC_END__+0x18>
6388                	ld	a0,0(a5)
0002b303          	ld	t1,0(t0)
406503f7          	add16	t2,a0,t1
b9638593          	addi	a1,t2,-1130
00b03533          	snez	a0,a1
8082                	ret
```
And I tried that:
```c
int main()
{
  uint64_t a = 0x12345, b = 0x54321;
  return __rv__add16(a, b) != add16(b, a);
}
```
And it was compiled out (`li a0,0; ret`).
So machine descriptions allow compiler to know that this is similar code in some cases.
What should I do if I want my generic code to be compiled with p-ext insns' emitting?

linsinan1995 commented:
Thanks for your feedbacks and helps!

There are two ways to generate p-ext instructions:
1. manually call intrinsic functions, such as `__rv__add16` and `__rv__v_sadd16`(for vector type).
2. automatically generated by GCC (not the case for unspec instructions, see riscv.md "unspec" enum).

Take add16 as an example, you can see `add16` from these three functions, and I think the third one is what you are looking for.

```c
#include <rvp_intrinsic.h>
#include <stdlib.h>
#include <stdint.h>

// typedef short int16x4_t __attribute__((vector_size (8)));
// defined in rvp_intrinsic.h

static __attribute__ ((noinline))
int16x4_t v_sadd16_intrinsic (int16x4_t ra, int16x4_t rb)
{
  return __rv__v_sadd16 (ra, rb);
}

static __attribute__ ((noinline))
int64_t sadd16_intrinsic (int64_t ra, int64_t rb)
{
  return __rv__add16 (ra, rb);
}

// v4hi type addition can generate add16 instruction
static __attribute__ ((noinline))
int16x4_t v_sadd16_spn (int16x4_t ra, int16x4_t rb)
{
  return ra + rb;
}
```

./obj-rv64gcp/bin/riscv64-unknown-elf-gcc -S -O0 add16.c
```
        .file   "add16.c"
        .option nopic
        .attribute arch, "rv64i2p0_m2p0_a2p0_f2p0_d2p0_c0p0_p2p0_zpn2p0_zpsf2p0"
        .attribute unaligned_access, 0
        .attribute stack_align, 16
        .text
        .align  1
        .type   v_sadd16_intrinsic, @function
v_sadd16_intrinsic:
        addi    sp,sp,-32
        sd      s0,24(sp)
        addi    s0,sp,32
        sd      a0,-24(s0)
        sd      a1,-32(s0)
        ld      a4,-24(s0)
        ld      a5,-32(s0)
        add16   a5, a4, a5
        mv      a0,a5
        ld      s0,24(sp)
        addi    sp,sp,32
        jr      ra
        .size   v_sadd16_intrinsic, .-v_sadd16_intrinsic
        .align  1
        .type   v_sadd16_spn, @function
v_sadd16_spn:
        addi    sp,sp,-32
        sd      s0,24(sp)
        addi    s0,sp,32
        sd      a0,-24(s0)
        sd      a1,-32(s0)
        ld      a4,-24(s0)
        ld      a5,-32(s0)
        add16   a5, a4, a5
        mv      a0,a5
        ld      s0,24(sp)
        addi    sp,sp,32
        jr      ra
        .size   v_sadd16_spn, .-v_sadd16_spn
        .align  1
        .type   sadd16_intrinsic, @function
sadd16_intrinsic:
        addi    sp,sp,-32
        sd      s0,24(sp)
        addi    s0,sp,32
        sd      a0,-24(s0)
        sd      a1,-32(s0)
        ld      a4,-24(s0)
        ld      a5,-32(s0)
        add16   a5, a4, a5
        mv      a0,a5
        ld      s0,24(sp)
        addi    sp,sp,32
        jr      ra
        .size   sadd16_intrinsic, .-sadd16_intrinsic
        .ident  "GCC: (GNU) 10.2.0"
```

marcfedorow commented on gcc/common/config/riscv/riscv-common.c:
```diff
@@ -51,25 +51,27 @@ struct riscv_implied_info_t
+  {"zks", "zkg", 0},
+  {"zks", "zkb", 0},
+  {"p", "zpn", 0},
+  {"p", "zprv", 64},
```
I am not sure if xlen should be added here.
Look for example at Zknh from k-ext.
It has 4 sha256* insns for both rv32 and rv64, 6 sha512* rv32-only insns and 4 sha512* rv64-only insns.
If we do follow this logic, Zprv just have an empty quantity of rv32-suitable insns (i.e. all its insns are rv64+-only).
This actually rises a question if Zprv is needed at all -- I asked it [here](https://github.com/riscv/riscv-p-spec/issues/76).

marcfedorow commented on gcc/common/config/riscv/riscv-common.c:
```diff
@@ -642,8 +644,9 @@ riscv_subset_list::handle_implied_ext (const char *ext,
       /* TODO: Implied extension might use different version.  */
-      add (implied_info->implied_ext, major_version, minor_version,
-	   explicit_version_p);
+      if (xlen () >= implied_info->xlen)
```
If this logic will not be reverted, it should check either equality or if this xlen is available (in case of rv32-only Zeds).
```c
if (!implied_info->xlen // any xlen
    || (xlen() & implied_info->xlen))
```

marcfedorow replied:
By "available" I meant that `implied_info->xlen[_requirement]` is a mask of legal xlens (e.g. `16 | 32` passes this check on xlen=16 and xlen=32 but not xlen=64) -- this is why I suggested a bit-wise check.
Anyway, everything will blow up with any xlen not in [32, 64] so it should be fine.

linsinan1995 replied:
I agree. but since the xlen checking for instructions in [riscv-binutils](https://github.com/riscv/riscv-binutils-gdb/blob/riscv-binutils-2.37/opcodes/riscv-dis.c#L507) works in this way, so I think maybe we can just leave it there.

marcfedorow commented on gcc/common/config/riscv/riscv-common.c:
```diff
@@ -51,25 +51,27 @@ struct riscv_implied_info_t
+  {"p", "zpn", 0},
+  {"p", "zprv", 64},
+  {"p", "zpsf", 0},
   {NULL, NULL}
```
If this logic stays, I expect `{NULL, NULL, NULL}` or `{NULL, NULL, 0}` here.
It expands to zeros anyway by the power of C, yet it seems to be 'right' to explicitly place this zero.

marcfedorow commented on gcc/config/riscv/rvp.md:
```diff
@@ -0,0 +1,7058 @@
+}
+[(set_attr "type" "dsp")])
+
+(define_insn "kdm64_internal"
```
Wrong descriptions.
Check with
```
int main() {
    return 0
        || (__rv__kdmbb(0x80008000, 0x80008000) - 0x7FFFFFFF)
        || (__rv__kdmbt(0x80008000, 0x80008000) - 0x7FFFFFFF)
        || (__rv__kdmtt(0x80008000, 0x80008000) - 0x7FFFFFFF)
    ;
}
```

linsinan1995 replied:
Thanks, and actually `kdm64_internal` was unused for intrinsics.

I just added a patch for it to replace the previous unspec patterns, and I will update an execution version of test cases, and I think it would be helpful for checking the correctness. Thank you again.

marcfedorow commented on gcc/config/riscv/rvp.md:
```diff
@@ -0,0 +1,7058 @@
+(define_insn "kslra<VECI:mode><X:mode>"
+  [(set (match_operand:VECI 0 "register_operand"                  "=r")
+	(if_then_else:VECI
+	  (lt:X (match_operand:X 2 "register_operand"             " r")
```
Incorrect check.
If my register contains 0x00000010, it is a negative value in terms of kslra16 but obviously a positive value in general meaning (i.e. gcc thinks that it `lt X, 0` if false).

marcfedorow commented on gcc/config/riscv/rvp.md:
```diff
@@ -0,0 +1,7058 @@
+  [(set_attr "type" "simd")
+   (set_attr "mode" "V4HI")])
+
+;; S|U MUL16
```
```c
return 0xFF797D61F26D4629 != __rv__smul16(0x01234567, 0x89ABCDEF);
```
Seems like this puts 31..0 to 63..32 and some extension to 63..32.
This returned 0xc0800000ff797d61, does not seem as a correct answer.

marcfedorow commented on gcc/config/riscv/rvp.md:
```diff
@@ -0,0 +1,7058 @@
+  [(set (match_operand:V2SI 0 "register_operand"                             "=r")
+	(mult:V2SI (any_extend:V2SI (match_operand:V2HI 1 "register_operand" "%r"))
+		   (any_extend:V2SI (match_operand:V2HI 2 "register_operand" " r"))))]
+  "TARGET_ZPSF && !TARGET_64BIT"
```
`return 0xFF797D61F26D4629 != __rv__smul16(0x01234567, 0x89ABCDEF);`

marcfedorow commented on gcc/config/riscv/rvp.md:
```diff
@@ -0,0 +1,7058 @@
+  [(set_attr "type"   "dsp")
+   (set_attr "mode"   "DI")])
+
+(define_insn "sraiu"
```
```c
int main()
{

  if (0xFFFF1358 != __rv__sra_u(0x89ABCDEF, 0x0F)) return 1;
  if (0xFFFFFFFF != __rv__sra_u(0x89ABCDEF, 0x1F)) return 2;
  if (0x00000000 != __rv__sra_u(0xCDEF0000, 0x1F)) return 3;
  return 0;
```
returns 3
```asm
<main>:
	89abd7b7          	lui	a5,0x89abd
	7745                	lui	a4,0xffff1
	def78293          	addi	t0,a5,-529 # <__text_end+0x9abac1f>
	35870313          	addi	t1,a4,856 # <__text_end+0x7ffef188>
	4505                	li	a0,1
	d4f296f7          	srai.u	a3,t0,0xf
	00669a63          	bne	a3,t1,<main+0x28>
	53fd                	li	t2,-1
	d5f29577          	srai.u	a0,t0,0x1f
	00751463          	bne	a0,t2,<main+0x26>
	450d                	li	a0,3
	8082                	ret
	4509                	li	a0,2
	8082                	ret
```

linsinan1995 commented:
FYI, this update only changed the author email.

linsinan1995 committed:
[MD] add inst: KMADS32, KMADRS32, KMAXDS32, KMSDA32, KMSXDA32

linsinan1995 committed:
[Builtin] add intrinsic builtins

linsinan1995 committed:
[intrinsic header] Add interface for intrinsics

linsinan1995 committed:
[testcases] add testcases

linsinan1995 commented:
update to 0.9.10

linsinan1995 committed:
[MD] add inst: fsr, fsri, fsrw, rev

linsinan1995 committed:
[MD] add move pattern for vector mode

marcfedorow commented on gcc/config/riscv/rvp.md:
```diff
@@ -0,0 +1,6987 @@
+   (set_attr "mode" "DI")])
+
+;; uk|k add|sub w|h
+(define_int_attr karth_insn
```
U?K(sub|add)[HW] could be described similar to U?K(sub|add)<bits> + sign extension

linsinan1995 replied:
> U?K(sub|add)[HW] could be described similar to U?K(sub|add) + sign extension

Hi @marcfedorow , this is added in the latest update. Thanks for your feedback.

linsinan1995 committed:
rvp: fix wrong format in maddr32 and msubr32

linsinan1995 committed:
rvp: use gcc standard name for rv32p mac operation

linsinan1995 committed:
update testcases

linsinan1995 committed:
rvp: update to spec 0.9.11

0. update impl according to the change in 0.9.11
1. fix add/sub crash in rv32g
2. use standard name for saturation insn, e.g. kaddw and kaddh

manasabatchu commented:
while doing the ./configure --prefix what is the arch and abi

manasabatchu commented:
Assembler messages:
Error: rv64gc_zpn_zpsf: unknown prefixed ISA extension `zpn'
builtin-rvp64-add16.s:3: Error: rv64i2p0_m2p0_a2p0_f2p0_d2p0_c2p0_zpn2p0_zpsf2p0: unknown prefixed ISA extension `zpn'
builtin-rvp64-add16.s:21: Error: unrecognized opcode `add16 a5,a4,a5'
builtin-rvp64-add16.s:41: Error: unrecognized opcode `add16 a5,a4,a5'
builtin-rvp64-add16.s:61: Error: unrecognized opcode `add16 a5,a4,a5'
how to solve this problem?

pz9115 commented:
> Assembler messages: Error: rv64gc_zpn_zpsf: unknown prefixed ISA extension `zpn' builtin-rvp64-add16.s:3: Error: rv64i2p0_m2p0_a2p0_f2p0_d2p0_c2p0_zpn2p0_zpsf2p0: unknown prefixed ISA extension `zpn' builtin-rvp64-add16.s:21: Error: unrecognized opcode `add16 a5,a4,a5' builtin-rvp64-add16.s:41: Error: unrecognized opcode `add16 a5,a4,a5' builtin-rvp64-add16.s:61: Error: unrecognized opcode `add16 a5,a4,a5' how to solve this problem?

I think you need also use Binutils that supported P-ext,

We had update p-ext toolchain support with p-spec v0.9.11, here is the latest repo:

https://github.com/plctlab/riscv-gcc/tree/riscv-gcc-p-ext
https://github.com/plctlab/riscv-binutils-gdb/tree/riscv-binutils-p-ext

You can find some build record refer in this issue:
https://github.com/riscv-collab/riscv-gnu-toolchain/issues/1298#issuecomment-1651755805

manasabatchu commented:
I think you need also use Binutils that supported P-ext,

We had update p-ext toolchain support with p-spec v0.9.11, here is the latest repo:

https://github.com/plctlab/riscv-gcc/tree/riscv-gcc-p-ext
https://github.com/plctlab/riscv-binutils-gdb/tree/riscv-binutils-p-ext

You can find some build record refer in this issue:
https://github.com/riscv-collab/riscv-gnu-toolchain/issues/1298#issuecomment-1651755805
 I installed new version toolchain but I am getting this error 
Assembler messages:
Error: rv32imafdc_zpn: unknown prefixed ISA extension `zpn'
/tmp/ccKYHrO7.s:3: Error: rv32i2p0_m2p0_a2p0_f2p0_d2p0_c2p0_zpn2p0: unknown prefixed ISA extension `zpn'
/tmp/ccKYHrO7.s:11: Error: unrecognized opcode `add16 a0,a0,a2'
/tmp/ccKYHrO7.s:12: Error: unrecognized opcode `add16 a1,a1,a3'

manasabatchu commented:
> > Assembler messages: Error: rv64gc_zpn_zpsf: unknown prefixed ISA extension `zpn' builtin-rvp64-add16.s:3: Error: rv64i2p0_m2p0_a2p0_f2p0_d2p0_c2p0_zpn2p0_zpsf2p0: unknown prefixed ISA extension `zpn' builtin-rvp64-add16.s:21: Error: unrecognized opcode `add16 a5,a4,a5' builtin-rvp64-add16.s:41: Error: unrecognized opcode `add16 a5,a4,a5' builtin-rvp64-add16.s:61: Error: unrecognized opcode `add16 a5,a4,a5' how to solve this problem?
> 
> I think you need also use Binutils that supported P-ext,
> 
> We had update p-ext toolchain support with p-spec v0.9.11, here is the latest repo:
> 
> https://github.com/plctlab/riscv-gcc/tree/riscv-gcc-p-ext https://github.com/plctlab/riscv-binutils-gdb/tree/riscv-binutils-p-ext
> 
> You can find some build record refer in this issue: [riscv-collab/riscv-gnu-toolchain#1298 (comment)](https://github.com/riscv-collab/riscv-gnu-toolchain/issues/1298#issuecomment-1651755805)

Thank you for helping, now it's working fine

manasabatchu commented:
While doing compilation I am getting this error
/home/manasa/PSIMD/riscv-gnu-toolchain/binutils/opcodes/riscv-opc.c:1819:20: error: ‘INSN_CLASS_ZPN’ undeclared here (not in a function); did you mean ‘INSN_CLASS_ZKNH’?
 {"add8",        0, INSN_CLASS_ZPN, "d,s,t",     MATCH_ADD8, MASK_ADD8, match_opcode, 0 },
                    ^~~~~~~~~~~~~~
                    INSN_CLASS_ZKNH
/home/manasa/PSIMD/riscv-gnu-toolchain/binutils/opcodes/riscv-opc.c:1821:21: error: ‘INSN_CLASS_ZPSF’ undeclared here (not in a function); did you mean ‘INSN_CLASS_ZPN’?
 {"add64",       32, INSN_CLASS_ZPSF, "nds_rdp,nds_rsp,nds_rtp", MATCH_ADD64, MASK_ADD64, match_opcode, 0 },
                     ^~~~~~~~~~~~~~~
                     INSN_CLASS_ZPN
/home/manasa/PSIMD/riscv-gnu-toolchain/binutils/opcodes/riscv-opc.c:1830:20: error: ‘INSN_CLASS_ZBPBO’ undeclared here (not in a function); did you mean ‘INSN_CLASS_ZBKB’?
 {"clz32",       32,INSN_CLASS_ZBPBO,  "d,s",    MATCH_CLZ, MASK_CLZ, match_opcode, INSN_ALIAS },
                    ^~~~~~~~~~~~~~~~
                    INSN_CLASS_ZBKB
/home/manasa/PSIMD/riscv-gnu-toolchain/binutils/opcodes/riscv-opc.c:2009:61: error: ‘MATCH_SHAMT_REV8_H’ undeclared here (not in a function); did you mean ‘MATCH_SHAMT_REV8_32’?
 {"swap8",       0, INSN_CLASS_ZBPBO, "d,s",     MATCH_GREVI|MATCH_SHAMT_REV8_H, MASK_GREVI|(OP_MASK_SHAMT << 20), match_opcode, INSN_ALIAS },
                                                             ^~~~~~~~~~~~~~~~~~
                                                             MATCH_SHAMT_REV8_32
/home/manasa/PSIMD/riscv-gnu-toolchain/binutils/opcodes/riscv-opc.c:2029:49: error: ‘MATCH_UKSTSA16’ undeclared here (not in a function); did you mean ‘MATCH_URSTSA16’?
 {"ukstsa16",    0, INSN_CLASS_ZPN, "d,s,t",     MATCH_UKSTSA16, MASK_UKSTSA16, match_opcode, 0 },
                                                 ^~~~~~~~~~~~~~
                                                 MATCH_URSTSA16
/home/manasa/PSIMD/riscv-gnu-toolchain/binutils/opcodes/riscv-opc.c:2029:65: error: ‘MASK_UKSTSA16’ undeclared here (not in a function); did you mean ‘MASK_URSTSA16’?
 {"ukstsa16",    0, INSN_CLASS_ZPN, "d,s,t",     MATCH_UKSTSA16, MASK_UKSTSA16, match_opcode, 0 },
                                                                 ^~~~~~~~~~~~~
                                                                 MASK_URSTSA16
/home/manasa/PSIMD/riscv-gnu-toolchain/binutils/opcodes/riscv-opc.c:2065:49: error: ‘MATCH_RDOV’ undeclared here (not in a function); did you mean ‘MATCH_ROL’?
 {"rdov",        0, INSN_CLASS_ZPN, "d",         MATCH_RDOV, MASK_RDOV, match_opcode, INSN_ALIAS },
                                                 ^~~~~~~~~~
                                                 MATCH_ROL
/home/manasa/PSIMD/riscv-gnu-toolchain/binutils/opcodes/riscv-opc.c:2065:61: error: ‘MASK_RDOV’ undeclared here (not in a function); did you mean ‘MASK_DIV’?
 {"rdov",        0, INSN_CLASS_ZPN, "d",         MATCH_RDOV, MASK_RDOV, match_opcode, INSN_ALIAS },
                                                             ^~~~~~~~~
                                                             MASK_DIV
/home/manasa/PSIMD/riscv-gnu-toolchain/binutils/opcodes/riscv-opc.c:2066:49: error: ‘MATCH_CLROV’ undeclared here (not in a function); did you mean ‘MATCH_CLRS8’?
 {"clrov",       0, INSN_CLASS_ZPN, "",          MATCH_CLROV, 0xffffffffU, match_opcode, INSN_ALIAS },
how to solve this problem?

manasabatchu commented:
Hi I am getting this error while doing compilation process 
./.././gcc/gcc/config/riscv/rvp.md:21:50: error: unknown mode `V4QI'
../.././gcc/gcc/config/riscv/rvp.md:21:73: note: following context is ` (V2HI "!TARGET_64BIT")'
Makefile:2572: recipe for target 's-preds-h' failed
can you help me how to solve this error

pz9115 commented:
> Hi I am getting this error while doing compilation process ./.././gcc/gcc/config/riscv/rvp.md:21:50: error: unknown mode `V4QI' ../.././gcc/gcc/config/riscv/rvp.md:21:73: note: following context is ` (V2HI "!TARGET_64BIT")' Makefile:2572: recipe for target 's-preds-h' failed can you help me how to solve this error

Can you support the source file for me to reproduce this problem，thanks.

pz9115 commented:
> Hi I am getting this error while doing compilation process ./.././gcc/gcc/config/riscv/rvp.md:21:50: error: unknown mode `V4QI' ../.././gcc/gcc/config/riscv/rvp.md:21:73: note: following context is ` (V2HI "!TARGET_64BIT")' Makefile:2572: recipe for target 's-preds-h' failed can you help me how to solve this error

Can you support the source file for me to reproduce this problem

manasabatchu commented:
Hi @pz9115, this is the source code file [link](https://github.com/plctlab/riscv-gcc/blob/riscv-gcc-p-ext/gcc/config/riscv/rvp.md) which I'm consuming in my project. I have already stated the triggered error line in my previous comment, please let me know if you need any other information.

manasabatchu commented:
Hi @pz9115, you have any idea why I am getting this error and what are the files I want change for resolving this error .

manasabatchu commented:
Hi @pz9115  after doing the modifications in gcc and binutils folder andI am doing  make i getting this error can you help me how to solve this problem 
make[5]: Leaving directory '/home/manasa/PSIMD/riscv-gnu-toolchain/build-newlib/riscv32-unknown-elf/newlib'
Making all in libc
make[5]: Entering directory '/home/manasa/PSIMD/riscv-gnu-toolchain/build-newlib/riscv32-unknown-elf/newlib/libc'
Making all in argz
make[6]: Entering directory '/home/manasa/PSIMD/riscv-gnu-toolchain/build-newlib/riscv32-unknown-elf/newlib/libc/argz'
riscv32-unknown-elf-gcc -B/home/manasa/PSIMD/riscv-gnu-toolchain/build-newlib/riscv32-unknown-elf/newlib/ -isystem /home/manasa/PSIMD/riscv-gnu-toolchain/build-newlib/riscv32-unknown-elf/newlib/targ-include -isystem /home/manasa/PSIMD/riscv-gnu-toolchain/newlib/newlib/libc/include -B/home/manasa/PSIMD/riscv-gnu-toolchain/build-newlib/riscv32-unknown-elf/libgloss/riscv32 -L/home/manasa/PSIMD/riscv-gnu-toolchain/build-newlib/riscv32-unknown-elf/libgloss/libnosys -L/home/manasa/PSIMD/riscv-gnu-toolchain/newlib/libgloss/riscv32    -DPACKAGE_NAME=\"newlib\" -DPACKAGE_TARNAME=\"newlib\" -DPACKAGE_VERSION=\"4.1.0\" -DPACKAGE_STRING=\"newlib\ 4.1.0\" -DPACKAGE_BUGREPORT=\"\" -DPACKAGE_URL=\"\" -I. -I/home/manasa/PSIMD/riscv-gnu-toolchain/newlib/newlib/libc/argz -fno-builtin -DHAVE_NANOSLEEP      -O2 -D_POSIX_MODE -ffunction-sections -fdata-sections    -mcmodel=medlow -c -o lib_a-dummy.o `test -f 'dummy.c' || echo '/home/manasa/PSIMD/riscv-gnu-toolchain/newlib/newlib/libc/argz/'`dummy.c
Assembler messages:
Error: cannot find default versions of the ISA extension `p'
Error: rv32imaf_d0p0p_zbpbo_zpn_zpsf: unknown prefixed ISA extension `zbpbo'
Error: internal: bad RISC-V opcode (unknown operand type `nds_rdp,nds_rsp,nds_rtp'): add64 nds_rdp,nds_rsp,nds_rtp
Fatal error: internal: broken assembler.  No assembly attempted
Makefile:393: recipe for target 'lib_a-dummy.o' failed
make[6]: *** [lib_a-dummy.o] Error 1
make[6]: Leaving directory '/home/manasa/PSIMD/riscv-gnu-toolchain/build-newlib/riscv32-unknown-elf/newlib/libc/argz'
Making all in stdlib
And you have any document for adding new extension to riscv-gnu-toolchain can please share to me

manasabatchu commented:
Hi when  I am running the p extension testcase I am getting this error how to solve it?

/home/manasa/PSIMD/riscv-gnu-toolchain/lib/gcc/riscv32-unknown-elf/10.2.0/include/rvp_intrinsic.h:311:1: error: incompatible types when returning type 'int' but 'uint8xN_t' {aka 'uint8x4_t'} was expected
  311 | CREATE_RVP_INTRINSIC_VECTOR (uint8xN_t, uadd8, uint8xN_t, uint8xN_t)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~
Makefile:19: recipe for target 'add_8.riscv' failed
make: *** [add_8.riscv] Error 1
